mathext/ell_carlson.go

```diff
@@ -0,0 +1,154 @@
+// Copyright ©2017 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package mathext
+
+import (
+	"math"
+)
+
+// EllipticRF computes the symmetric elliptic integral R_F(x,y,z):
+//  R_F(x,y,z) = (1/2)\int_{0}^{\infty}{1/s(t)} dt,
+//  s(t) = \sqrt{(t+x)(t+y)(t+z)}.
+//
+// The arguments x, y, z must satisfy the following conditions, otherwise the function returns math.NaN():
+//  0 ≤ x,y,z ≤ upper,
+//  lower ≤ x+y,y+z,z+x,
+// where:
+//  lower = 5/(2^1022) = 1.112536929253601e-307,
+//  upper = (2^1022)/5 = 8.988465674311580e+306.
+//
+// The definition of the symmetric elliptic integral R_F can be found in NIST Digital Library of Mathematical Functions
+// (http://dlmf.nist.gov/19.16.E1). The original Fortran code was published as Algorithm 577 in ACM TOMS (http://doi.org/10.1145/355958.355970).
+// This code is also available as a part of SLATEC Common Mathematical Library (http://netlib.org/slatec/index.html). Later, Carlson described
+// an improved version in http://dx.doi.org/10.1007/BF02198293 (also available at https://arxiv.org/abs/math/9409227).
+func EllipticRF(x, y, z float64) float64 {
+	const (
+		lower = 5.0 / (1 << 256) / (1 << 256) / (1 << 256) / (1 << 254) // 5*2^-1022
+		upper = 1 / lower
+		tol   = 1.2674918778210762260320167734407048051023273568443e-02 // (3ε)^(1/8)
+	)
+	if x < 0 || y < 0 || z < 0 || math.IsNaN(x) || math.IsNaN(y) || math.IsNaN(z) {
+		return math.NaN()
+	}
+	if upper < x || upper < y || upper < z {
+		return math.NaN()
+	}
+	if x+y < lower || y+z < lower || z+x < lower {
+		return math.NaN()
+	}
+
+	A0 := (x + y + z) / 3
+	An := A0
+	Q := math.Max(math.Max(math.Abs(A0-x), math.Abs(A0-y)), math.Abs(A0-z)) / tol
+	xn, yn, zn := x, y, z
+	mul := 1.0
+
+	for Q >= mul*math.Abs(An) {
+		xnsqrt, ynsqrt, znsqrt := math.Sqrt(xn), math.Sqrt(yn), math.Sqrt(zn)
+		lambda := xnsqrt*ynsqrt + ynsqrt*znsqrt + znsqrt*xnsqrt
+		An = (An + lambda) * 0.25
+		xn = (xn + lambda) * 0.25
+		yn = (yn + lambda) * 0.25
+		zn = (zn + lambda) * 0.25
+		mul *= 4
+	}
+
+	X := (A0 - x) / (mul * An)
+	Y := (A0 - y) / (mul * An)
+	Z := -(X + Y)
+	E2 := X*Y - Z*Z
+	E3 := X * Y * Z
+
+	// http://dlmf.nist.gov/19.36.E1
+	return (E3*(6930*E3+E2*(15015*E2-16380)+17160) + E2*((10010-5775*E2)*E2-24024) + 240240) / (240240 * math.Sqrt(An))
```

Comment:
I don't think these large multiplications are a good idea, because of floating point precision reasons. I would write it as the formula is written. 
```
math.Sqrt(An) * (1 - 0.1 * E2 + 1.0/14 * E3 + 1.0/24* E2*E2 - 3.0/44 * E2 * E3)
```

I also don't see the utility in the citation comment. The same formula is given in Carlson.